src/allmydata/testing/web.py

```diff
@@ -0,0 +1,225 @@
+
+    def add_data(self, kind, data):
+        return self._uri.add_data(kind, data)
+
+
+@attr.s
+class _FakeCapability(object):
+    """
+    """
+    data=attr.ib()
+
+
+KNOWN_CAPABILITIES = [
+    getattr(allmydata.uri, t).BASE_STRING
+    for t in dir(allmydata.uri)
+    if hasattr(getattr(allmydata.uri, t), 'BASE_STRING')
+]
+
+
+def capability_generator(kind):
+    """
+    :param str kind: the kind of capability, like `URI:CHK`
+
+    :returns: a generator that yields new capablities of a particular
+        kind.
+    """
+    if kind not in KNOWN_CAPABILITIES:
+        raise ValueError(
+            "Unknown capability kind '{} (valid are {})'".format(
+                kind,
+                ", ".join(KNOWN_CAPABILITIES),
+            )
+        )
+    # what we do here is to start with empty hashers for the key and
+    # ueb_hash and repeatedly feed() them a zero byte on each
+    # iteration .. so the same sequence of capabilities will always be
+    # produced. We could add a seed= argument if we wanted to produce
+    # different sequences.
+    number = 0
+    key_hasher = hashlib.new("sha256")
+    ueb_hasher = hashlib.new("sha256")
+
+    # capabilities are "prefix:<128-bits-base32>:<256-bits-base32>:N:K:size"
+    while True:
+        number += 1
+        key_hasher.update("\x00")
+        ueb_hasher.update("\x00")
+
+        key = base32.b2a(key_hasher.digest()[:16])  # key is 16 bytes
+        ueb_hash = base32.b2a(ueb_hasher.digest())  # ueb hash is 32 bytes
+
+        cap = u"{kind}:{key}:{ueb_hash}:{n}:{k}:{size}".format(
+            kind=kind,
+            key=key,
+            ueb_hash=ueb_hash,
+            n=1,
+            k=1,
+            size=number * 1000,
+        )
+        yield cap.encode("ascii")
+
+
+class _FakeTahoeUriHandler(Resource, object):
+    """
+    """
+
+    isLeaf = True
+    _data = None
+    _capability_generators = None
+
+    def _generate_capability(self, kind):
+        """
+        :param str kind: any valid capability-string type
+
+        :returns: the next capability-string for the given kind
+        """
+        if self._capability_generators is None:
+            self._capability_generators = dict()
+
+        if kind not in self._capability_generators:
+            self._capability_generators[kind] = capability_generator(kind)
+        capability = next(self._capability_generators[kind])
+        return capability
+
+    def add_data(self, kind, data):
+        """
+        adds some data to our grid
+
+        :returns: a capability-string
+        """
+        assert isinstance(data, bytes)
+
+        cap = self._generate_capability(kind)
+        if self._data is None:
+            self._data = dict()
+        assert cap not in self._data, "already have '{}'".format(cap)
+        self._data[cap] = data
+        return cap
+
+    def render_PUT(self, request):
+        data = request.content.read()
+        return self.add_data("URI:CHK:", data)
```

Comment:
I guess this should do some error handling and return the same kind of failure as Tahoe-LAFS really does - INTERNAL ERROR?  CONFLICT?  And the reachable error case should be tested.  Not sure about the unreachable error case (unknown cap type ... `_FakeTahoeUriHandler` could have a direct unit test for that case I guess, but what behavior is coherent?  It's a new failure mode that Tahoe-LAFS itself can't have, I think?)

Reply:
So, I did add some code for "overwrite" versus not .. but "in general" am trying to limit how much gets added in one PR, and magic-folder doesn't (yet?) care about overwrites